libraries/app.rb

```diff
@@ -0,0 +1,190 @@
+          directory install_dir do
+            recursive true
+            action :delete
+          end if new_resource.force && exist?
+
+          if new_resource.url.end_with?('dmg')
+            install_dmg
+          elsif new_resource.url.end_with?('xip')
+=begin
+            temp_pkg_dir = ::File.join(Chef::Config[:file_cache_path], "Xcode_#{new_resource.version}")
+            temp_pkg_file = ::File.join(temp_pkg_dir, ::File.basename(new_resource.url))
+
+            directory temp_pkg_dir do
+              recursive true
+            end
+
+# For some reason this doesn't work with some XIPs, Apple?
+            remote_file "Xcode" do
+              source new_resource.url
+              checksum new_resource.checksum
+              path temp_pkg_file
+              action :create
+            end
+
+            bash 'Extracting xip using xar' do
+              user 'root'
+              cwd temp_pkg_dir
+              code <<-EOH
+                #!/bin/sh
+                xar -xf #{temp_pkg_file}
+              EOH
+            end
+
+            ark "Xcode" do
+              name 'Content'
+              url "file://#{temp_pkg_dir}/Content"
+              extension 'tar.gz'
+              path '/Applications'
+              action :put
+            end
+            directory temp_pkg_dir do
+              recursive true
+              action :delete
+              only_if { Dir.exist?(temp_pkg_dir) }
+            end
+=end
+          else
+            raise 'Unsupported package provided Xcode installer must be provided as a DMG'
+          end
+
+          post_install
+          accept_eula
+          delete_installer_file if exist?
+        end
+
+        def install_dmg
+          dmg_package new_resource.app do
+            source new_resource.url
+            checksum new_resource.checksum
+            dmg_name ::File.basename(new_resource.url, '.dmg')
+            owner 'root'
+            type 'app'
+            action :install
+          end
+        end
+
+        def post_install
+          ruby_block 'Xcode post install' do
+            block do
+              if new_resource.multi_install || !new_resource.install_root.eql?('/applications')
+                ::File.rename('/Applications/Xcode.app', install_dir)
+              end
+
+              # Install any additional packages hiding in the Xcode installation path
+              xcode_packges =
+                ::Dir.entries("#{install_dir}/Contents/Resources/Packages/").select { |f| !::File.directory? f }
+              xcode_packges.each do |pkg|
+                execute "Installing Xcode package [#{pkg}]" do
+                  command "sudo installer -pkg #{install_dir}/Contents/Resources/Packages/#{pkg} -target /"
+                end
+              end unless xcode_packges.nil?
+
+              execute 'xcode-select' do
+                command "xcode-select -s #{install_dir}/Contents/Developer"
```

Comment:
The `post_install` function is only called if the function 'install_app' is called.